### DISCUSSION TITLE
Snapshot Tests - Mock Store

### DISCUSSION THREAD

**Discussion by TiagoSantosSilva** [pointfreeco/swift-composable-architecture#2135] [Q&A] [answered]
Hi folks,

I'm currently trying to add snapshot tests to one of my views. I'm having some issues with the effects that are triggered when the snapshot runs and the `.onAppear/.task` is called on the view.

My current structure is the following

View 

```swift
struct HomeView: View {
    let store: StoreOf<HomeFeature>
    @ObservedObject var viewStore: ViewStoreOf<HomeFeature> 

    init(store: StoreOf<HomePageFeature>) {
        self.store = store
        viewStore = ViewStore(store)
}

var body: some View {
    Group {
        if let error = viewStore.state.error {
            ErrorView(...)
        } else {
            homeView
        }
    }
    .onAppear {
        viewStore.send(.onAppear)
    }
    .isLoading(viewStore.state.isLoading)
}
```

The `isLoading` modifier is a custom modifier we have to display a shimmering view to the user.

Store

```swift
struct HomeFeature: ReducerProtocol {
    enum Action: Equatable {
        case onAppear
        ...
        ...
    }

    struct State: Equatable {
        var isLoading: Bool = false
        var error: EquatableError?
    }

    var body: some ReducerProtocol<State, Action> {
        Reduce { state, action in

            let effect: EffectTask<Action>

            switch action {
            case .onAppear:
                state.isLoading = true
                
               (... some long running effect e.g: network call)
           }
     }
}
```

I'm then trying to run a snapshot test for when the `State` contains an error (and is not loading), it's currently like this:

```swift
func testErrorState() {
    let nsError = NSError(
            domain: "test",
            code: 0, userInfo: [
                NSLocalizedDescriptionKey: "Error message"
            ]
        )
    let error = EquatableError(nsError)
    let store = StoreOf<HomeFeature>(initialState: .init(error: error), reducer: HomeFeature())
    let sut = HomePageView(store: mockedStore(error: error))
        .frame(width: 390, height: 850)
    assertViewSnapshot(sut)
}
```

The issue, however, is that since `.onAppear` is called, I'm not being able to make it so that the error message is displayed since `isLoading` is being set to `true` and that has priority over the error.

Is there a way to make it so that either:

• The Store I'm initializing the `HomeView` with has no actions/its actions do nothing?
• The View I'm initializing/testing doesn't get `onAppear` triggered?

I've tried both using the `withDependencies { ... }` initializer so that my mocked service would return an error right away in order to try to achieve the error state that way but it also doesn't seem to work here.
Using `.actionless` seems to provide a `Store` that I'm not able to use on the initializer 😕.

Ha anyone faced a similar issue and was able to find a solution for it?

Thanks!

**Comment by mbrandonw** [comment, marked as answer]
You can create your store with no reducer to represent a completely inert store that does nothing when you send actions to it. That allows you to craft whatever you state you want and just hand it to the view.

This is how it can look:

```swift
let store = StoreOf<HomeFeature>(initialState: .init(error: error), reducer: EmptyReducer())
```

Or if you are using the newest version of TCA that allows specifying the reducer in a trailing builder closure:

```swift
let store = StoreOf<HomeFeature>(initialState: .init(error: error)) {
}
```

**Reply by TiagoSantosSilva** [reply]
Thanks for the help Brandon.

We're not using the newest version of TCA yet but the first solution worked like a charm.

Cheers!
